### PR TITLE
Pass "record-wrapped" types by 'const in' after resolution.

### PR DESCRIPTION
**Pull request by vasslitvinov**
A.k.a. somebody tell me why we were passing them by reference?

This change reaffirms that after resolution the values of
record-wrapped types, i.e. _array, _domain, _distribution,
are treated as records -- not as arrays/domains/distributions
that they implement.

Therefore these records are always passed around by value,
i.e. 'const in'. In the generated code this corresponds to
passing pointers to implementation classes, e.g. BlockArr,
DefaultRectangularDom, etc.

These records and pointers are always constant. That's because,
in Chapel, once an array/... variable is created, it is associated
always with the same array/..., i.e. the same instance of BaseArr/...

Before this change, these records/pointers were themselves passed
by reference, in a few cases. This inflates the size of the AST
and the generated code and may reduce optimization opportunities
for the C compiler.

IMPLEMENTATION DETAILS

* These functions now return 'const in' for record-wrapped types
after resolution:

  constIntentForType()
  blankIntentForType()

Also resolveIntents() switches the existing 'ref' and 'const ref'
intents for record-wrapped types to 'const in'.

* Additionally, flattenFunctions.cpp/passByRef() returns 'false'
for symbols of record-wrapped types.

This comes handy when the function being flattened has an outer variable
of a record-wrapped type, such as a temp. We make sure it is passed
by value as well.

Note that the former checks sym->hasFlag(FLAG_ARRAY) et al.
never fired I think, because these flags are on the type symbols,
not the argument/variable symbols.

* Notes on remoteValueForwarding.cpp/isSufficientlyConst():

** The following assert:

```
  if (isRecordWrappedType(argvaltype))
    INT_ASSERT(arg->intent == INTENT_CONST_IN  &&
               !arg->type->symbol->hasFlag(FLAG_REF));
```

mostly documents the current state. If it fails, we are probaly
missing some pass-by-value opportunity somewhere.

** The isSufficientlyConst() function itself is now useless.
Indeed, RVF is done only when arg's type is non-FLAG_REF.
In these cases isSufficientlyConst() always returns false.

I am leaving this function around in case we start RVF
non-REF things.

* I had to tweak checkResolved.cpp/isDefinedAllPaths(),
which counts the number of definitions of 'ret'.
For arrays, a "definition" used to come from passing 'ret'
into =() by 'ref' intent. Now the code is the same,
except the intent is 'const in'.
(Recall that checkResolved comes after resolution and
resolveIntents.)
So I have isDefinedAllPaths() accept 'const in'
for record-wrapped types as a "definition" as well.

This loses precision, because we cannot tell passing
an array by 'ref' from 'const ref'. So potentially
we can now create AST where an array 'ret' is left
uninitialized, yet passed by 'const ref' into something.
I accepted this in order to keep the semantics straight.

Alternatively we could do our 'ref'->'const in' conversion
after checkResolved. I chose against this because there is
really no good place to put such conversion: checkResolved
is followed immediately by processIteratorYields.
The conversion itself would be very simple:

```
  forv_Vec(ArgSymbol, arg, gArgSymbols) {
    if (isRecordWrappedType(arg->type) &&
        (arg->intent & INTENT_FLAG_REF))
      // enact low-level semantics for domain, array, or distribution
    arg->intent = INTENT_CONST_IN;
  }
```

ALSO

* There is a small improvement in comm counts in SSCA2.

### CODE REVIEW

File: compiler/passes/checkResolved.cpp [hunk 1]
```diff
@@ -139,6 +139,14 @@ isDefinedAllPaths(Expr* expr, Symbol* ret, RefSet& refs)
                arg->intent == INTENT_REF))
             return 1;
 
+          // For arrays/domains/distributions, we just converted all 'ref' args
```
Comment (hildeth) [inline compiler/passes/checkResolved.cpp]: When the new constructor story is fully-implemented, definition will occur when the variable is initialized.  In that case it is moot whether assignment is counted as a definition.  As I stated today, we (the compiler, that is) should not know anything more about the '=' operator than that it is a function.  It is entirely up to the implementing class whether it actually takes 2 arguments and what are the intents of those arguments.
So, I accept "const in" in connection with record-wrapped types as a workaround.  Eventually, the need for this workaround will go away.  "const in" seems right for arguments of that type, in any case.

File: test/studies/ssca2/performance/SSCA2_commDiags.good [hunk 2]
```diff
@@ -34,7 +34,7 @@ Starting Graph Generation in parallel
 max number of outgoing edges 12
 number of disconnected vertices 2
 
-Kernel 2: (get = 10, get_nb = 0, put = 7, put_nb = 0, test_nb = 0, wait_nb = 0, try_nb = 0, fork = 0, fork_fast = 0, fork_nb = 6) (get = 44, get_nb = 0, put = 0, put_nb = 0, test_nb = 0, wait_nb = 0, try_nb = 0, fork = 8, fork_fast = 0, fork_nb = 0) (get = 40, get_nb = 0, put = 0, put_nb = 0, test_nb = 0, wait_nb = 0, try_nb = 0, fork = 4, fork_fast = 0, fork_nb = 0) (get = 40, get_nb = 0, put = 0, put_nb = 0, test_nb = 0, wait_nb = 0, try_nb = 0, fork = 4, fork_fast = 0, fork_nb = 0)
```
Comment (hildeth) [inline test/studies/ssca2/performance/SSCA2_commDiags.good]: Very nice.  Keep the evil up.

**Comment by hildeth** [comment]
OK, looks fine to me.
Fine to go in as-is, though you might make a note in comments circa line 142 in checkResolved.cpp, about that particular special case being a workaround before initialization is done just right.

**Comment by bradcray** [comment]
Though I'm happy to see unnecessary references go away at codegen time, I'm not convinced that changing resolveIntents is the right way to go about it.  Specifically, recall that the goal of resolve intents was to determine the Chapel-source-level intents of arguments, not the C-codegen-level intents.  I'd argue that anything related to how to generate C (or LLVM or Fortran or ...) should be done closer to that point in compilation rather than immediately after resolution.  Specifically, if we succeeded in our goal of incorporating resolveIntents into resolution itself in order to remove all the de facto intent resolution that goes on there today and used this logic, we wouldn't be able to distinguish an array argument that the user labeled as 'const in' by the user vs. one that was passed by blank intent, even though the former should result in a deep copy of its elements and the latter not.  So, I would've proposed logic close to ccodegen (or in it) that said something like "we're generating C where arguments are passed by value, and this is a record wrap type that needs to be passed by reference according to Chapel, so using C's const in suffices" rather than labeling it as CONST_IN in the IR.  Are there reasons not to take that approach?

**Comment by vasslitvinov** [comment]
I absolutely agree that resolution should be done treating arrays as arrays, and intents should be Chapel-level intents. We are currently pretty close to this I think, and this PR does not change that.

Once we are out of resolution, specifically after checkResolved, we are entering a gray area where we treat wrapper records as records and struggle when we actually want to know something Chapel-level about the corresponding arrays.

In other words, TODAY your (Brad's) "closer to codegen" starts right after resolution. The testing results support this - NOTHING noticed the semantic change in intents, except one spot in checkResolved. Tom's work on reference counting of arrays via UMM of wrapper records is also based on the "compiler knows nothing about arrays" paradigm.

TOMORROW, i.e. after 1.11, I am very much in favor of doing inlining, constant propagation, LICM, escape analysis, etc., while treating arrays as arrays. This PR merely makes official what we already do today.

**Comment by bradcray** [comment]
Does that mean that tomorrow, i.e. after 1.11, you, Vass, will be willing to rewrite resolveIntents to mark these cases as "by-ref" again and to have any necessary logic that avoids turning them into references in C moved into codegen or appropriate passes closer to it;  even though I'd prefer for you to do that today?  (i.e., before feature freeze)  Rationalizing a change that isn't where we want to be headed by saying "nothing broke, so therefore it's OK" doesn't seem appropriate to me, particularly when it happens a few days before feature freeze and nobody knew about the effort or had an opportunity to weigh in on it other than a reviewer set up off-list.

I also don't see the basis of your argument that this patch preserves that the IR's intents reflect Chapel-level intents when passing an array by blank or ref intent showed up as "ref" in the --log= output last week but now shows up as "const in".

I think the current approach is problematic (as, it sounds like, you do) for any post-resolution optimizations that want to reason about arrays as arrays since it makes it impossible to tell whether the array was passed in by blank or const in or ref intent.  Whether or not any current passes rely on that array nature or not seems beside the point to me.

An alternative approach that I'd be much more supportive of would be to have both Chapel-level and C-level intents stored in ArgSymbols and to set both in resolveIntents so that downstream passes can query either.  But personally, even that approach would seem like it's binding to a particular code generation strategy much earlier than we should.

**Comment by vasslitvinov** [comment]
Brad, I am on your side for wanting a clear separation between Chapel-level and C-level semantics. I propose that after 1.11 we create a "lower intents" pass in the compiler that will serve as a clear line: before that is's Chapel and after that it's C.

Here is the reason why I did not do it: it takes more than just rewriting "ref" intents to "const in". For example, wrapper records are currently considered "isRecord()", and in the Chapel world that won't work. Also I suspect that we will want to insert copy constructors and assignment operators as part of such a pass.

Believe me, I did not change how the compiler treats arrays after resolution. I just made it clear. And yes, my intention was to make it _painfully_ clear, to motivate us to take action about it. To support my stand, prior to my change:

* Passing by 'ref', the way we treat it in the compiler, means passing the wrapper record by 'ref'. We can observe this in the generated code in some cases.

* Passing by 'in' means passing the corresponding array by 'ref'. I just compiled this little example:

```
  proc main {
    var A: [1..2] int;
    testme(A);
  }
  
  proc testme(in B) {
    writeln(B);
  }
```

In the IR at checkResolved(), 'B' has 'const in' intent - and then it's initCopy'ed into _formal_tmp_B. Meaning that 'B' is expected to reference the actual-argument array. This is without my change.


Yes I can rewrite resolveIntents as you request - let me know if you still want me to do it before feature freeze. This will amount to deleting 3 lines and adding 7 lines, plus the cost of creating a new compiler pass. Mind you, it will need to happen before flattenFunctions(), because it is there that we start passing wrapper records by reference. So I will convert processIteratorYields() into a super-pass instead of adding a new pass.

Instead, my suggestion - and request - is to implement a comprehensive solution, after 1.11.

**Comment by hildeth** [comment]
For my part, I approved of this pull request with the understanding that it did not change the semantics of array manipulations, but did remove an extra level of indirection that was being employed in the manipulation of the wrapper records that contain the array implementations.  After the removeWrapRecords pass, wrap-record instances are replaced by pointers or integers; it makes little sense to pass these around by reference, since passing them by value (i.e. copying the pointer or the integer itself) will always be more efficient.  If my understanding of this is incorrect on any point, it might be best to take a step back and consider backing out the change.
Because the resolveIntents pass relies on the resolved type of formal arguments to determine that argument-passing convention, it can be argued that the output of that pass is already dealing with low-level concepts.  If we are concerned about preserving the intent (both literally and figuratively) expressed in the user code, we should do exactly that: Rather than re-writing the intent tags in one or more passes, it might be better to leave them in the state they had immediately after parsing and instead use predicates (e.g. "isConstRefAfterResolution()") to interpret those intents.  This would provide a clear separation between the high-level and low-level views of the program without necessarily causing the high-level view to be lost at any particular point in the translation.
I believe there is time remaining to consider whether the patch should remain part of the release.  Let us meet and discuss this after Brad returns.

**Pull request merged**